File: apps/desktop/src/types/metrics.ts
export interface RequestRecord {
  id: number
  source: string
  session_id: string | null
  timestamp: string
  model: string | null
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  cost: number
  project: string | null
  latency_ms: number | null
}

export interface MetricsSummary {
  total_cost: number
  total_requests: number
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  cache_hit_rate: number
  session_count: number
}

export interface ProjectUsage {
  branch: string
  project: string
  source: string
  request_count: number
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  cost: number
  session_count: number
}

export interface TrendPoint {
  bucket: string
  source: string
  input_tokens: number
  output_tokens: number
  cost: number
}

export interface SessionSummary {
  session_id: string
  source: string
  model: string | null
  request_count: number
  input_tokens: number
  output_tokens: number
  cost: number
  start_time: string
  end_time: string
}

export interface ActiveSession {
  source: string
  session_id: string
  model: string | null
  last_active: string
}

export type TimeRange = 'today' | 'week' | 'month' | 'all'
export type SourceView = 'providers' | 'projects' | 'sessions'

File: apps/server/app/schemas/metrics.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime


class MetricsSummary(BaseModel):
    total_cost: float
    total_requests: int
    input_tokens: int
    output_tokens: int
    cache_read_tokens: int
    cache_write_tokens: int
    cache_hit_rate: float
    session_count: int


class TrendPoint(BaseModel):
    bucket: str
    source: str
    input_tokens: int
    output_tokens: int
    cost: float


class SessionSummary(BaseModel):
    session_id: str
    source: str
    model: Optional[str]
    request_count: int
    input_tokens: int
    output_tokens: int
    cost: float
    start_time: datetime
    end_time: datetime


class ActiveSession(BaseModel):
    source: str
    session_id: str
    model: Optional[str]
    last_active: datetime
